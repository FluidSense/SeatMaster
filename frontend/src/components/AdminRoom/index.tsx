import React, { ChangeEvent, Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router';
import { ThunkDispatch } from 'redux-thunk';
import { IPostRoom } from '../../API/interfaces';
import { IStore } from '../../store';
import { IRoom } from '../ViewRooms';
import {
  createRoomAction,
  deleteRoomAction,
  resetPage,
  updateRoomAction,
} from './actions';
import './adminRoom.css';
import { ROUTE_TO } from './constants';
import Presentational from './Presentational';

interface IState {
  buttonDisabled: boolean;
  redirect: boolean;
  room: IRoom;
  showAlert: boolean;
  modalOpen: boolean;
}

interface IProps {
  location: {
    room?: IRoom;
  };
}

interface IDispatchProps {
  createRoom: (data: IPostRoom) => void;
  updateRoom: (data: IPostRoom, id: number) => void;
  deleteRoom: (id: number) => void;
  reset: () => void;
}

interface IStateProps {
  submitted?: boolean;
  error?: string;
  room: any;
}

type Props = IDispatchProps & IStateProps & IProps;

// tslint:disable-next-line:class-name
class _Container extends Component<Props, IState> {
  constructor(props: Props) {
    super(props);
    this.state = {
      buttonDisabled: true,
      modalOpen: false,
      redirect: false,
      room: {
        id: -1,
        info: '',
        name: '',
        seats: {
          count: 0,
          seats: [],
        },
      },
      showAlert: false,
    };
  }

  public componentDidUpdate = (prevProps: Props, prevState: IState) => {
    const { room, showAlert } = this.state;
    const { submitted, reset } = this.props;
    // Display alertstripe
    if (submitted === false && prevProps.submitted === undefined) {
      this.setState({ showAlert: true });
      reset();
    }
    if (showAlert) setTimeout(() => this.setState({ showAlert: false }), 5000);
    // Disable or enable create room button
    if (prevState.room.name !== room.name || prevState.room.info !== room.info) {
      if (room.name.trim() !== '' && room.name.trim() !== '') {
        this.setState({ buttonDisabled: false });
      } else {
        this.setState({ buttonDisabled: true });
      }
    }
  }

  public componentDidMount = () => {
    const { room } = this.props.location;
    if (room) {
      this.setState({
        room,
      });
    }
  }

  public render() {
    const { buttonDisabled, room, showAlert, modalOpen } = this.state;
    const { submitted, reset, error } = this.props;
    const roomExists = room.id !== -1;
    const onClick = roomExists ? this.update : this.create;

    if (submitted) {
      reset();
      return <Redirect to={ROUTE_TO} />;
    }
    return (
      <Presentational
        alertMessage={error}
        buttonDisabled={buttonDisabled}
        deleteRoom={this.delete}
        onClick={onClick}
        room={room}
        roomExists={roomExists}
        setName={this.setName}
        setNotes={this.setNotes}
        showAlert={showAlert}
        toggleModal={this.toggleModal}
        modalOpen={modalOpen}
      />
    );
  }

  private setNotes = (notesEvent: ChangeEvent<HTMLInputElement>) => {
    const roomNotes = notesEvent.target.value;
    this.setState(prevState => ({
      room: {
        ...prevState.room,
        info: roomNotes,
      },
    }));
  }

  private setName = (nameEvent: ChangeEvent<HTMLInputElement>) => {
    const roomName = nameEvent.target.value;
    this.setState(prevState => ({
      room: {
        ...prevState.room,
        name: roomName,
      },
    }));
  }

  private create = () => {
    const { info: roomNotes, name: roomName } = this.state.room;
    const { createRoom } = this.props;
    const body = { info: roomNotes, name: roomName };
    createRoom(body);
  }

  private toggleModal = () => {
    this.setState({ modalOpen: !this.state.modalOpen });
  }

  private delete = () => {
    const { room } = this.state;
    const { deleteRoom } = this.props;
    if (!room) return null;
    deleteRoom(room.id);
  }

  private update = () => {
    const { room } = this.state;
    const { info: roomNotes, name: roomName } = room;
    const { updateRoom } = this.props;
    if (!room) return null;
    const body = { info: roomNotes, name: roomName };
    updateRoom(body, room.id);
  }
}

const mapStateToProps = (state: IStore) => ({
  error: state.adminRoom.error,
  room: state.adminRoom.room,
  submitted: state.adminRoom.success,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<{}, {}, any>) => ({
  createRoom: (data: IPostRoom) => dispatch(createRoomAction(data)),
  deleteRoom: (id: number) => dispatch(deleteRoomAction(id)),
  reset: () => dispatch(resetPage()),
  updateRoom: (data: IPostRoom, id: number) => dispatch(updateRoomAction(data, id)),
});

const AdminRoom = connect(
  mapStateToProps,
  mapDispatchToProps,
)(_Container);

export default AdminRoom;
