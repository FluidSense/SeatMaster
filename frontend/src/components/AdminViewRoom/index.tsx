import { connect } from 'react-redux';
import { IStore } from '../../store';
import Presentational from './Presentational';

const mapStateToProps = (state: IStore) => ({
  applications: state.applications.applications,
});

const Container = connect(
  mapStateToProps,
)(Presentational);

export default Container;
