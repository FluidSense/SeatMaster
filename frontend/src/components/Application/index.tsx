import { Sidetittel } from 'nav-frontend-typografi';
import * as React from 'react';
import { IUser } from '../../API/interfaces';
import { ISeat } from '../ViewRooms';
import Presentational from './Presentational';

export interface IPartnerApplication {
  comments?: string;
  id: number;
  needs?: string;
  preferredRoom?: string;
  seatRollover?: boolean;
  status: string;
  rank: string;
  user: IUser;
}

export interface IApplication extends IPartnerApplication {
  partnerApplication?: IPartnerApplication;
  seat?: ISeat;
}

interface IState {
  modalIsOpen: boolean;
}

const _TITLE = 'My application';

export class Application extends React.Component<{}, IState> {
  constructor(props: object) {
    super(props);
    this.state = {
      modalIsOpen: false,
    };
  }

  public render() {
    return (
      <div className="main-content">
        <Sidetittel className="page-title">{_TITLE}</Sidetittel>
        <Presentational modalIsOpen={this.state.modalIsOpen} changeModal={this.changeModal} />
      </div >
    );
  }

  private changeModal = (modalOpen: boolean) => { this.setState({ modalIsOpen: modalOpen }); };
}

export default Application;
