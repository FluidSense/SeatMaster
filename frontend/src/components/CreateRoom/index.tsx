import React, { ChangeEvent, Component } from 'react';
import { Redirect } from 'react-router';
import { postRoom } from '../../API/calls';
import {
  POST_HEADERS,
  POST_NEW_ROOM_URL,
  POST_ROOM_NAME,
  POST_ROOM_NOTES,
} from '../commonConstants';
import { IRoom } from '../ViewRooms';
import Presentational from './Presentational';

interface IState {
  roomName: string;
  roomNotes: string;
  buttonDisabled: boolean;
  showAlert: boolean;
  redirect: boolean;
  room?: IRoom;
}

interface IProps {
  match: {
    params: {
      id: number;
    };
  };
}

// tslint:disable-next-line:class-name
class CreateRoom extends Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);
    this.state = {
      buttonDisabled: true,
      redirect: false,
      roomName: '',
      roomNotes: '',
      showAlert: false,
    };
  }

  public componentDidUpdate = (prevProps: {}, prevState: IState) => {
    const { roomName, roomNotes, showAlert } = this.state;
    // Display alertstripe
    if (showAlert && !prevState.showAlert) {
      setTimeout(() => this.setState({ showAlert: false }), 5000);
    }
    // Disable or enable create room button
    if (prevState.roomName !== roomName || prevState.roomNotes !== roomNotes) {
      if (roomName.trim() !== '' && roomNotes.trim() !== '') {
        this.setState({ buttonDisabled: false });
      } else {
        this.setState({ buttonDisabled: true });
      }
    }
  }

  public componentDidMount = () => {
    const { id } = this.props.match.params;
    fetch(`http://localhost:5000/room/${id}`)
      .then(response => response.json())
      .then(room => this.setState({ room, roomName: room.name, roomNotes: room.info }));
  }

  public render() {
    const { id } = this.props.match.params;
    const { buttonDisabled, roomName, roomNotes, showAlert, redirect } = this.state;
    const onClick = id ? this.updateRoom : this.createRoom;
    if (redirect) return <Redirect to={'/admin/rooms'} />;
    return (
      <Presentational
        roomName={roomName}
        roomNotes={roomNotes}
        buttonDisabled={buttonDisabled}
        setNotes={this.setNotes}
        setName={this.setName}
        onClick={onClick}
        deleteRoom={this.deleteRoom}
        showAlert={showAlert}
        roomId={id}
      />
    );
  }

  private setNotes = (notesEvent: ChangeEvent<HTMLInputElement>) => {
    const roomNotes = notesEvent.target.value;
    this.setState({ roomNotes });
  }
  private setName = (nameEvent: ChangeEvent<HTMLInputElement>) => {
    const roomName = nameEvent.target.value;
    this.setState({ roomName });
  }

  private createRoom = () => {
    const { roomName, roomNotes } = this.state;
    const body = { [POST_ROOM_NOTES]: roomNotes, [POST_ROOM_NAME]: roomName };
    postRoom(body)
      .then(() => this.setState({ redirect: true }), () => this.setState({ showAlert: true }));
  }

  private deleteRoom = () => {
    const { room } = this.state;
    if (!room) return null;
    fetch(`http://localhost:5000/room/${room.id}`, {
      headers: POST_HEADERS,
      method: 'delete',
    })
      .then(response => response.status)
      .then((statusCode) => {
        if (statusCode === 200) this.setState({ redirect: true });
        if (statusCode === 400) this.setState({ showAlert: true });
      });
  }

  private updateRoom = () => {
    const { roomName, roomNotes, room } = this.state;
    if (!room) return null;
    const body = { info: roomNotes, name: roomName };
    fetch(`http://localhost:5000/room/${room.id}`, {
      body: JSON.stringify(body),
      headers: POST_HEADERS,
      method: 'put',
    })
      .then(response => response.status)
      .then((statusCode) => {
        if (statusCode === 200) this.setState({ redirect: true });
        if (statusCode === 400) this.setState({ showAlert: true });
      });
  }
}

export default CreateRoom;
