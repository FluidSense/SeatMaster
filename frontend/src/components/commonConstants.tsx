// Etikett types
export const ETIKETT_FOCUS = 'fokus';
export const ETIKETT_INFO = 'info';
export const ETIKETT_SUCCESS = 'suksess';
export const ETIKETT_WARNING = 'advarsel';

export const POST_ROOM_NAME = 'name';
export const POST_ROOM_NOTES = 'info';
export const POST_HEADERS = { 'Content-Type': 'application/json' };
// Visual date time format for end users
export const DATE_FORMAT_INPUT_FIELD = 'DD/MM/YYYY - H:mm';

// URL for database
export const POST_NEW_SEASON_URL = 'http://localhost:5000/season/createSeason';
export const POST_NEW_ROOM_URL = 'http://localhost:5000/room/createRoom';
export const GET_USER_BY_NAME_URL = 'http://localhost:5000/user/';
export const POST_NEW_USER_URL = 'http://localhost:5000/user/createUser';

// actions
export const SET_USER_DATA = 'SET_USER_DATA';

// Application status codes
// These are mirrored from the backend
export const APP_NOT_FOUND = 'NOT_FOUND';
export const APP_SUBMITTED = 'SUBMITTED';
export const APP_DENIED = 'DENIED';
export const APP_APPROVED = 'APPROVED';
