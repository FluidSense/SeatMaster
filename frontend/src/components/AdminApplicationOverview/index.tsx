import React from 'react';
import { connect } from 'react-redux';
import { AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { IStore } from '../../store';
import { IApplication } from '../Application';
import { fetchAllApplications } from './actions';
import './adminApplicationOverview.css';
import Presentational from './Presentational';

interface IStateProps {
  applications: IApplication[];
}

interface IDispatchProps {
  getAllApplications: () => ThunkAction<void, {}, {}, AnyAction>;
}

type Props = IStateProps & IDispatchProps;

// tslint:disable-next-line:class-name
class _Container extends React.Component<Props> {
  public componentDidMount() {
    this.props.getAllApplications();
  }

  public render() {
    return (
      <Presentational
        applications={this.props.applications}
      />);
  }
}

const mapStateToProps = (state: IStore) => ({
  applications: state.applications.applications,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<{}, {}, any>) => ({
  getAllApplications: () => dispatch(fetchAllApplications()),
});

const Container = connect(
  mapStateToProps,
  mapDispatchToProps,
)(_Container);

export default Container;
