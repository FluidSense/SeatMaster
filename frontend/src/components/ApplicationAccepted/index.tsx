import React from 'react';
import { IApplication } from '../Application';
import { IRoom } from '../ViewRooms';
import SeatDisplay from './SeatDisplay';

interface IProps {
  application: IApplication;
  fetchRoomInfo: (id: number) => void;
  rooms: IRoom[];
}

const ApplicationAccepted: React.FunctionComponent<IProps> = (props) => {
  const { rooms, fetchRoomInfo } = props;
  const { seat } = props.application;

  if (!seat) return null;
  const seatedRoom = rooms.filter(room => room.id === seat.roomId);
  if (!seatedRoom.length) {
    fetchRoomInfo(seat.roomId);
    return null;
  }
  return (
    <>
    <h1>Your seat</h1>
    <SeatDisplay seat={seat} room={seatedRoom[0]}/>
    </>
  );
};

export default ApplicationAccepted;
