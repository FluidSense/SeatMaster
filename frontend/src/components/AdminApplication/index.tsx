import React, { Component } from 'react';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import { IPostAdminApplicationForm } from '../../API/interfaces';
import { IStore } from '../../store';
import { fetchAllApplications } from '../AdminApplicationOverview/actions';
import { IApplication } from '../Application';
import { removeStudent } from '../AssignSeat/actions';
import { APP_NOT_FOUND } from '../commonConstants';
import { updateSingleApplication } from '../EditApplication/actions';
import LoadingPageSpinner from '../LoadingPageSpinner';
import Page404 from '../Page404';
import { IRoom, ISeat } from '../ViewRooms';
import { fetchAllRooms } from '../ViewRooms/actions';
import { resetPageStatus } from './actions';
import Presentational from './Presentational';

interface IStateProps {
  rooms: IRoom[];
  seatInfo?: ISeat;
  modalOpen: boolean;
  applications: IApplication[];
  match: {
    params: {
      id: string;
    };
  };
}

interface IDispatchProps {
  removeStudentFromSeat: (roomId: number, seatId: string) => void;
  fetchApplications: () => void;
  fetchRooms: () => void;
  resetStatus: () => void;
  updateApplication: (id: number, app: IPostAdminApplicationForm) => void;
}

interface IState {
  fetchedApplications: boolean;
  fetchedRooms: boolean;
}

type Props = IStateProps & IDispatchProps;

// tslint:disable-next-line:variable-name
class AdminApplication extends Component<Props, IState> {
  constructor(props: Props) {
    super(props);
    this.state = {
      fetchedApplications: false,
      fetchedRooms: false,
    };
  }

  public componentDidMount() {
    if (!this.props.applications.length && !this.state.fetchedApplications) {
      this.setState({ fetchedApplications: true });
      this.props.fetchApplications();
    }
    if (!this.props.rooms.length && !this.state.fetchedRooms) {
      this.setState({ fetchedRooms: true });
      this.props.fetchRooms();
    }
  }

  public render() {
    const { removeStudentFromSeat, applications, rooms, updateApplication } = this.props;
    const { id } = this.props.match.params;
    const application = applications.find(app => app.id === Number(id));
    if (!(applications.length && rooms) || !rooms.length) return <LoadingPageSpinner />;
    if (!application) return <Page404 />;

    return (
       <Presentational
        application={application}
        rooms={rooms}
        removeStudentFromSeat={removeStudentFromSeat}
        updateApplication={updateApplication}
       />
    );
  }
}

const mapStateToProps = (state: IStore) => ({
  applications: state.applications.applications,
  rooms: state.rooms.rooms,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<{}, {}, any>) => ({
  fetchApplications: () => dispatch(fetchAllApplications()),
  fetchRooms: () => dispatch(fetchAllRooms()),
  removeStudentFromSeat: (roomId: number, seatId: string) =>
    dispatch(removeStudent(roomId, seatId)),
  resetStatus: () => dispatch(resetPageStatus()),
  updateApplication: (id: number, app: IPostAdminApplicationForm) => {
    return dispatch(updateSingleApplication(id, app));
  },
});

const Container = connect(
  mapStateToProps,
  mapDispatchToProps,
)(AdminApplication);

export default Container;
