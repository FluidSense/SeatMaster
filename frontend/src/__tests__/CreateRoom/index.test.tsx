import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import fetchMock from 'fetch-mock';
import * as React from 'react';
import { POST_NEW_ROOM_URL } from '../../components/commonConstants';
import CreateRoom from '../../components/CreateRoom/index';

describe('Create room functions', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<CreateRoom />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Updates roomName state through onchange', () => {
    const component = shallow(<CreateRoom />);
    const name = 'X-Wing';
    const inputName = component.dive().find('#input-room-name');
    inputName.simulate('change', { target: { value: name } });
    expect(component.state('roomName')).toEqual(name);
  });

  it('Updates roomNotes state through onchange', () => {
    const component = shallow(<CreateRoom />);
    const notes = 'These are some notes, bro';
    const inputNotes = component.dive().find('#input-room-notes');
    inputNotes.simulate('change', { target: { value: notes } });
    expect(component.state('roomNotes')).toEqual(notes);
  });

  it('Fetches with correct body', () => {
    fetchMock.mock(POST_NEW_ROOM_URL, 201);
    const component = shallow(<CreateRoom />);
    const submitButton = component.dive().find('#create-room-button');
    const newState = {
      ...component.state(),
      roomName: 'X-Wing',
      roomNotes: 'Notes',
    };
    component.setState(newState);
    const jsonBody = JSON.stringify({
      info: component.state('roomNotes'),
      name: component.state('roomName'),
    });
    submitButton.simulate('click');
    const fetchOptions = fetchMock.lastOptions();
    expect(fetchOptions.body).toEqual(jsonBody);
  });
});

describe('Create Room life cycle', () => {
  it('Update state and check if button is enabled', () => {
    const wrapper = shallow(<CreateRoom />) as any;
    const newState = {
      ...wrapper.state(),
      roomName: 'X-Wing',
      roomNotes: 'Notes',
    };
    wrapper.setState(newState);
    expect(wrapper.state('buttonDisabled')).toEqual(false);
  });
});
