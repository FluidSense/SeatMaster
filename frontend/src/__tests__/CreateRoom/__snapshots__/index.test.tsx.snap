// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create room functions renders correctly 1`] = `
<Presentational
  buttonDisabled={true}
  createRoom={[Function]}
  roomName=""
  roomNotes=""
  setName={[Function]}
  setNotes={[Function]}
  showAlert={false}
/>
`;
